### PR TITLE
Privacy Metadata initial

### PR DESCRIPTION
**Pull request by sebwrede**
This pull request adds privacy constraints as metadata. The privacy constraints are propagated from input to result of matrix multiplications. The privacy constraints currently only contain a simple boolean, which represents whether the data may be shared or not, but more complex constraints will later be added to the PrivacyConstraint class.

This pull request replaces the [Privacy Metadata pull request](https://github.com/apache/systemml/pull/894).

### CODE REVIEW

File: src/main/java/org/tugraz/sysds/parser/DataExpression.java [hunk 1]
```diff
@@ -109,31 +113,70 @@
 	
 	public static final String DELIM_SPARSE = "sparse";  // applicable only for write
 	
-	public static final String[] RAND_VALID_PARAM_NAMES = 
-		{RAND_ROWS, RAND_COLS, RAND_DIMS, RAND_MIN, RAND_MAX, RAND_SPARSITY, RAND_SEED, RAND_PDF, RAND_LAMBDA};
+	public static final Set<String> RAND_VALID_PARAM_NAMES =
```
Comment (j143) [inline src/main/java/org/tugraz/sysds/parser/DataExpression.java:116]: If this is a formatting change, the progress might be slowed. The best method would be to make any formatting changes in a different PR. or adhering to the existing formatting.

File: src/main/java/org/tugraz/sysds/runtime/instructions/Instruction.java [hunk 2]
```diff
@@ -27,6 +27,7 @@
 import org.tugraz.sysds.lops.Lop;
 import org.tugraz.sysds.parser.DataIdentifier;
```
Comment (j143) [inline src/main/java/org/tugraz/sysds/runtime/instructions/Instruction.java:28]: Can the class names be updated to reflect `org.apache`, instead of `org.tugraz` to trigger integration tests.

File: src/main/java/org/tugraz/sysds/runtime/privacy/PrivacyConstraint.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.tugraz.sysds.runtime.privacy;
+
+/**
```
Comment (j143) [inline src/main/java/org/tugraz/sysds/runtime/privacy/PrivacyConstraint.java:22]: While the design is fresh in memory, shall we add the implementation related documentation or design in this file.
Perhaps something like [this](https://github.com/apache/systemml/blob/cf74661016928e3413d693f939a67964f3256b19/src/main/java/org/apache/sysds/runtime/instructions/gpu/context/GPUMemoryManager.java#L62)

File: src/main/java/org/tugraz/sysds/parser/DataExpression.java (hunk 1 above)
Comment (sebwrede) [reply src/main/java/org/tugraz/sysds/parser/DataExpression.java:116]: I needed to add privacy to the parameter array and then I noticed inconsistencies in how parameters are checked using the arrays, so I decided to change it. It could easily be a separate PR and I will try to be better at making this kind of update in separate PRs in the future.

**Comment by sebwrede** [comment]
I will rebase this branch to resolve the conflicts. Please wait with any edits.

**Comment by sebwrede** [comment]
The rebase is done. The checks are failing because [Github's servers are having problems](https://www.githubstatus.com/).

**Comment by mboehm7** [comment]
LGTM - thanks for the initial integration @sebwrede. During the merge I made only some minor formatting fixes (tabs over spaces, verbosity) and slightly changed the hop/lop metadata handover.
